Use proper duck typing for JsProxy

hoodmane opened the pull request:
This resolves #1162. I use `PyType_FromSpecWithBases` to dynamically create the type of each `JsProxy` with only the appropriate slots. 

`PyType_FromSpecWithBases` seems to have had some backwards incompatible changes between Python version 3.8 and Python 3.9 so this may require some adjustment when we get to upgrading to version 3.9.

rth reviewed (commented):
This is very nice! 

A few comments below.
However I think we need more test,
 - for the cases when an exception is explicitly raised, we should check that it's indeed the case
 - in general I don't think that the conditional existence of the appropriate methods is currently tested for all of the below cases,
   ```
   #define IS_ITERABLE  (1<<0)
   #define IS_ITERATOR  (1<<1)
   #define HAS_LENGTH   (1<<2)
   #define HAS_GET      (1<<3)
   #define HAS_SET      (1<<4)
   #define HAS_HAS      (1<<5)
   #define HAS_INCLUDES (1<<6)
   #define IS_AWAITABLE (1<<7)
   #define IS_BUFFER    (1<<8)
   #define IS_CALLABLE  (1<<9)
   #define IS_ARRAY     (1<<10)
   ```
   ?

hoodmane commented on src/core/jsproxy.c:
```diff
@@ -286,6 +430,12 @@ JsProxy_Dir(PyObject* self, PyObject* _args)
+  if (hiwire_is_array(GET_JSREF(self))) {
+    // See comment about Array.keys in GetAttr
+    keys_str = PyUnicode_FromString("keys");
+    FAIL_IF_NULL(keys_str);
```
I think I specifically looked in the definition of `PyUnicode_FromString` and found at least one failing branch. Probably `MemoryError`, anything that can allocate can throw out of memory. Thing is I'm pretty sure that `PyUnicode_FromString` interns the string and it's basically unimaginable that `keys` won't be interned before running out of memory. So... yeah it probably can't fail.

hoodmane commented:
Thanks for the review @rth!

hoodmane commented on src/core/jsproxy.c:
```diff
@@ -222,13 +259,84 @@ JsProxy_length(PyObject* o)
+    return NULL;
+  }
+  PyErr_Format(PyExc_TypeError,
+               "list indices must be integers or slices, not %.200s",
```
This is copied verbatim from the CPython source. I don't know what it means either. Note that Python doesn't use the standard `printf` formatters, it uses [`PyUnicode_FromFormat`](https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_FromFormat) which is similar but slightly different in not fully documented ways.

Here is where I copied this from:
https://github.com/python/cpython/blob/213c155a460b8dd9e43901e4d61aa088cbac4221/Objects/listobject.c#L2856

hoodmane replied:
Presumably this truncates the length of the string to 200 if you have an extremely long type name?

hoodmane replied:
I found the relevant CPython comment:

>      *** WARNING ***
>
>       The format should limit the total size of the formatted output string to
>       1000 bytes.  In particular, this means that no unrestricted "%s" formats
>       should occur; these should be limited using "%.<N>s where <N> is a
>       decimal number calculated so that <N> plus the maximum size of other
>       formatted text does not exceed 1000 bytes.  Also watch out for "%f",
>       which can print hundreds of digits for very large numbers.

https://github.com/python/cpython/blob/e89380765df8f0f02c90ad417e164d1597bd0b05/Python/sysmodule.c#L3243

hoodmane commented:
@rth I think I addressed your comments aside from your request for more tests?

hoodmane commented:
I think `hiwire_get_method` is a bad name, it sounds like it gets a method, not like it calls a method called `get`. Not sure what the best alternative is though.

rth commented on src/core/hiwire.h:
```diff
@@ -413,6 +422,33 @@ hiwire_get_length(JsRef idobj);
+hiwire_has_get_method(JsRef idobj);
+
+JsRef
+hiwire_get_method(JsRef idobj, JsRef idkey);
```
Could you add a docstring here as well, as done below? At least one sentence saying what it does, as indeed the names can be confusing.

rth commented:
> I think hiwire_get_method is a bad name, it sounds like it gets a method, not like it calls a method called get

Maybe `hiwire_call_get_method` to be consistent with `hiwire_has_get_method` ?

hoodmane commented on src/tests/test_jsproxy.py:
```diff
@@ -508,3 +502,205 @@ def test_register_jsmodule_docs_example(selenium):
+        window.l[6] = 7;
+        await pyodide.runPythonAsync(`
+            from js import l
+            from pytest import raises
```
Would be nice to have a shim since `pytest.raises` functions like both `assertRaises` and `assertRaisesRegex`. But I don't think that `testing.py` is a good place for it: as much as possible I want tests to run when copied into the browser console in `console.html`. For now I can just duplicate it a bunch.

hoodmane commented:
Okay I think I addressed the recent comments. I updated most of the docstrings in `hiwire.c` to include a bit more information about what they do in various error conditions. It's sort of unfortunate that `EM_JS` coerces return values it doesn't know how to deal with to zero...

rth reviewed (approved):
Thanks! Generally LGTM but it's a lot of code to review.
